heroku_connect/models.py

```diff
@@ -0,0 +1,342 @@
+
+    .. _Trigger Log in Heroku Connect docs:
+        https://devcenter.heroku.com/articles/writing-data-to-salesforce-with-heroku-connect#understanding-the-trigger-log
+    .. _error handling:
+        https://devcenter.heroku.com/articles/writing-data-to-salesforce-with-heroku-connect#write-errors
+
+    """
+
+    # read-only fields
+    id = models.BigIntegerField(primary_key=True, editable=False)
+    created_at = models.DateTimeField(editable=False, null=True)
+    updated_at = models.DateTimeField(editable=False, null=True)
+    processed_at = models.DateTimeField(editable=False, null=True)
+    table_name = models.CharField(max_length=128, editable=False)
+    record_id = models.BigIntegerField(editable=False)
+    sf_id = models.CharField(max_length=18, editable=False, null=True, db_column='sfid')
+    action = models.CharField(max_length=7, editable=False, choices=TriggerLogAction.choices())
+    sf_message = models.TextField(editable=False, null=True, blank=True)
+
+    # TODO: these are more useful as HStoreFields (if 'django.contrib.postgres' in INSTALLED_APPS)
+    # leave them as textfields for now
+    _values = models.TextField(editable=False, null=True, blank=True, db_column='values')
+    _old = models.TextField(editable=False, null=True, blank=True, db_column='old')
+
+    # editable fields
+    state = models.CharField(max_length=8, null=False, blank=False,
+                             choices=TriggerLogState.choices())
+
+    objects = models.Manager.from_queryset(TriggerLogQuerySet)()
+
+    class Meta:
+        abstract = True
+        managed = False
+        get_latest_by = 'created_at'
+        ordering = ('id',)
+
+    is_archived = False
+
+    @classmethod
+    def capture_insert_from_model(cls, table_name, record_id, *, exclude_fields=()):
+        """Create a fresh insert record from the current model state in the database.
+
+        For read-write connected models, this will lead to the attempted creation of a
+        corresponding object in Salesforce.
+
+        Args:
+            table_name: The name of the table backing the connected model (without schema)
+            record_id: The primary id of the connected model
+            exclude_fields: The names of fields that will not be included in the write record
+
+        Raises:
+            LookupError: if ``table_name`` does not belong to a connected model
+
+        """
+        exclude_cols = ()
+        if exclude_fields:
+            model_cls = hc_models.registry.get_class_for_table_name(table_name)
+            exclude_cols = cls._fieldnames_to_colnames(model_cls, exclude_fields)
+
+        sql = """
+            SELECT "{schema}".hc_capture_insert_from_row(
+              hstore({schema}.{table_name}.*),
+              '{table_name}',
+              ARRAY[{exclude_cols}]::text[]
+            ) AS id
+            FROM "{schema}"."{table_name}"
+            WHERE id = %(record_id)s
+        """.format(
```

Comment:
If you want to write and your own sql I recommend using psycopg2's native formating and escaping functionality, see http://initd.org/psycopg/docs/sql.html